addon/components/frost-list-item-content.js

```diff
@@ -0,0 +1,101 @@
+import Ember from 'ember'
+const {get, isPresent} = Ember
+import computed, {readOnly} from 'ember-computed-decorators'
+import {Component} from 'ember-frost-core'
+import {PropTypes} from 'ember-prop-types'
+
+import layout from '../templates/components/frost-list-item-content'
+
+export default Component.extend({
+
+  // == Dependencies ==========================================================
+
+  // == Keyword Properties ====================================================
+  layout,
+
+  // == PropTypes =============================================================
+
+  propTypes: {
+    // Options - required
+    model: PropTypes.EmberObject.isRequired,
+    index: PropTypes.number.isRequired,
+    onExpand: PropTypes.func.isRequired,
+    onSelect: PropTypes.func.isRequired,
+
+    // Options - general
+    onSelectionChange: PropTypes.func,
+    itemTypes: PropTypes.object,
+    itemTypeKey: PropTypes.string,
+    isAnyTypedItemExpansion: PropTypes.bool,
+
+    // Options - sub-components
+    item: PropTypes.EmberComponent.isRequired,
+    itemExpansion: PropTypes.EmberComponent
+  },
+
+  getDefaultProps () {
+    return {
+      isAnyTypedItemExpansion: false
+    }
+  },
+
+  // == Computed Properties ===================================================
+
+  @readOnly
+  @computed('model', 'itemTypes', 'itemTypeKey')
+  _itemTypeContent (model, itemTypes, itemTypeKey) {
+    if (isPresent(model) && isPresent(itemTypes) && isPresent(itemTypeKey)) {
+      const itemType = model.get(itemTypeKey)
+
+      if (itemType in itemTypes) {
+        return get(itemTypes, itemType)
+      }
+    }
+  },
+
+  @readOnly
+  @computed('_itemTypeContent')
+  typedItemComponent (itemTypeContent) {
+    if (isPresent(itemTypeContent)) {
+      return get(itemTypeContent, 'item')
+    }
+  },
+
+  @readOnly
+  @computed('_itemTypeContent')
+  typedItemExpansionComponent (itemTypeContent) {
+    if (isPresent(itemTypeContent)) {
```

Comment:
You should rename your typedItemExpansionComponent to something like `_itemExpansion` and make it return either default `itemExpansion`, or typed `itemExpansion` or undefined. Then you can clean up a lot of boilerplate in your `frost-list-item-content` template.